packages/wallet/contract/src/smartWallet.js

```diff
@@ -1,90 +1,273 @@
-          ...assign(mutableState, { [key]: value }),
-        }),
-    });
-  });
+  // #region other state management
 
-  const marshaller = wallet.getMarshaller();
+  /**
+   * What we've been told by the Bank
+   *
+   * @type {WeakMapStore<Brand, BrandDescriptor>} */
+  const brandDescriptors = makeScalarWeakMapStore();
+  // #endregion
 
+  // #region publishing
+  // NB: state size must not grow monotonically
   const myWalletStorageNode = E(storageNode).makeChildNode(address);
-  const storedSubscriber = makeStoredSubscriber(
-    subscriber,
-    myWalletStorageNode,
+
+  /** @type {PublishKit<PursesRecord>} */
+  const pursePublishKit = makePublishKit();
+  const pursesSubscriber = makeStoredSubscriber(
+    pursePublishKit.subscriber,
+    E(myWalletStorageNode).makeChildNode('purses'),
+    marshaller,
+  );
+
+  const publishPurses = async () => {
+    const entries = Array.from(brandPurses.entries());
+    // indexed the same as brand array to merge
+    const balances = await Promise.all(
+      entries.map(([_, purse]) => E(purse).getCurrentAmount()),
+    );
+    const record = entries.map(([brand, _], i) => {
+      const desc = brandDescriptors.get(brand);
+      return { ...desc, purseBalance: balances[i] };
+    });
+    pursePublishKit.publisher.publish(record);
+  };
+
+  /** @type {PublishKit<OffersRecord>} */
+  const offerPublishKit = makePublishKit();
+  const offersSubscriber = makeStoredSubscriber(
+    offerPublishKit.subscriber,
+    E(myWalletStorageNode).makeChildNode('offers'),
     marshaller,
   );
+  /**
+   * @param {import('./offers.js').OfferStatus} status
+   */
+  const publishOfferStatus = status => {
+    const record = { [status.id]: status };
+    offerPublishKit.publisher.publish(record);
+  };
+  // #endregion
+
+  // #region issuer management
+  /** @type { (desc: BrandDescriptor) => Promise<void>} */
+  const addIssuer = async desc => {
+    // assert haven't received this issuer before.
+    const descriptorsHas = brandDescriptors.has(desc.brand);
+    const pursesHas = brandPurses.has(desc.brand);
+    assert(
+      !(descriptorsHas && pursesHas),
+      'repeated brand from bank asset subscription',
+    );
+    assert(
+      !(descriptorsHas || pursesHas),
+      'corrupted state; one store has brand already',
+    );
+
+    // make a purse for this issuer.
+    // first make the purse so that if anything fails we don't update stores
+    const purse = await E(desc.issuer).makeEmptyPurse();
+
+    // save all five of these in a collection (indexed by brand?) so that when
+    // it's time to take an offer description you know where to get the
+    // relevant purse. when it's time to make an offer, you know how to make
+    // payments. REMEMBER when doing that, need to handle every exception to
+    // put the money back in the purse if anything fails.
+    brandDescriptors.init(desc.brand, desc);
+    brandPurses.init(desc.brand, purse);
+
+    publishPurses();
```

Comment:
I expect we would publish an `{ issuer, brand, displayInfo, petname }` record here.

Reply:
That's basically `BrandDescriptor`, plus `displayInfo` to save a request.
https://github.com/Agoric/agoric-sdk/blob/c74df8148bdd198bef4e7fc2fec9209ddecdc89f/packages/wallet/contract/src/smartWallet.js#L28

I see now `getCurrentAmountNotifier` on purse. So we don't need `purseBalance`, but we do need to give clients a way to get that notifier. I'll include that in `PursesRecord` as `currentAmountNotifier`.